### PR TITLE
Install bookinfo application before graph test [OSSM-5723]

### PR DESCRIPTION
**Pull request by ferhoyos**
To load some known graph and perform more precise in cypress tests, `bookinfo` demo application is installed in Openshift before graph test is run. This test has been modified to check that `bookinfo` graph summary has been loaded correctly.

### CODE REVIEW

File: plugin/cypress/support/commands.ts
Comment (ferhoyos) [inline plugin/cypress/support/commands.ts:30]: To run cypress test in local CRC I have to clear login input before typing. This should not affect login for OCP instances.

**Comment by jmazzitelli** [comment]
I'm not crazy about duplicating all these hack scripts. We now have dups for download-istio, install-bookinfo-demo, wait-for-namespace .... not sure what others.

We change these hacks often, now we have to worry about changing them in multiple duplicate scripts.

Can we figure out a way to re-use? Even if its just "assume kiali/kiali is git cloned in a peer directory of kiali/openshift-servicemesh-plugin" so we can assume paths like "../../hack/" will work.

**Comment by ferhoyos** [comment]
@jmazzitelli yes I have thought on this... I have just duplicated the files because it is a bit urgent to complete OSSMC GA tests and answer what to do with duplicated files in other PR. Since they are 1:1 copies, maybe it is a good option to duplicate all hacks script from kiali/kiali into `hacks/kiali` folder (similar like we do with application code). What do you think? Is it feasible to extend our `copy-frontend-src-to-ossmc.sh` script to copy hack scripts too?

File: hack/istio/install-bookinfo-demo.sh
Comment (nrfox) [inline hack/istio/install-bookinfo-demo.sh:1]: Do we need to copy these scripts directly? If they change in `kiali/kiali` we'll need to remember to update them here as well. Can we reference them as a url instead to keep `kiali/kiali` the source of truth?

**Review by nrfox — CHANGES_REQUESTED** [review]

File: hack/istio/install-bookinfo-demo.sh
Comment (nrfox) [reply hack/istio/install-bookinfo-demo.sh:1]: Oops looks like mazz commented the same thing as me already...

Comment (nrfox) [reply hack/istio/install-bookinfo-demo.sh:1]: You could do something like `curl -sSL https://github.com/kiali/kiali/raw/master/hack/istio/install-bookinfo-demo.sh | bash`?

Comment (ferhoyos) [reply hack/istio/install-bookinfo-demo.sh:1]: But `install-bookinfo-demo.sh` makes calls to other hack scripts. Will it work if you only download that script?

Comment (ferhoyos) [reply hack/istio/install-bookinfo-demo.sh:1]: Notice that these scripts must be executed for OSSMC tests in Jenkins pipeline, so I am not sure if kiali/kiali git cloned in a peer directory will work for that use case....

Comment (jmazzitelli) [reply hack/istio/install-bookinfo-demo.sh:1]: Or... do a "git pull" or something like that to pull the hack/ directory down using https.

**Comment by ferhoyos** [comment]
I download just `hack/istio` folder from kiali repo using git sparse checkout method. @nrfox @jmazzitelli are you ok with this approach?

**Review by jmazzitelli — COMMENTED** [review]
Question: Is there a  reason you were not able to put the hack scripts under `_output`? I ask because that's the typical pattern in our kiali repos (the git-ignored and generated output is stored under the main `_output` directory). Is there a need to get these to go under the root "kiali" directory?

e.g. why not put them under `_output/kiali` and then call the scripts directly from there (`_output/kiali/hack/istio/download-istio.sh`)

As it is now, there is going to be this `kiali` directory that will look like its source but isn't really (at least it is git ignored so we won't accidentally check it in, but it does seem out of place).

**Comment by ferhoyos** [comment]
No, there is no reason to be under `kiali` folder. Your approach makes sense, I have put the Kiali scripts under `_output` folder.

**Review by jmazzitelli — DISMISSED** [review]
I haven't run this, but it looks OK to me.

File: plugin/cypress/support/step_definitions/hooks.ts
Comment (ScriptingShrimp) [inline plugin/cypress/support/step_definitions/hooks.ts:4]: Tests don't pass for me locally, bookinfo is not installed, will debug this for more detail.

I think this copypasta might be the issue, I will take a look if we can execute the bookinfo scripts more simpler than utilizing this hook from regular regression - that was maybe a bit over engineered approach 
```
   1) Graph page is displayed correctly
    ✓ Istio Config page is displayed correctly (6999ms)


  3 passing (1m)
  1 failing

  1) Kiali sidebar integration with OCP Console
       Graph page is displayed correctly:
     CypressError: `cy.exec('../_output/kiali/hack/istio/wait-for-namespace.sh -n bookinfo')` failed because the command exited with a non-zero code.

Pass `{failOnNonZeroExit: false}` to ignore exit code failures.

Information about the failure:
Code: 1

Stderr:
Error from server (NotFound): namespaces "kiali-operator" not found

https://on.cypress.io/exec
      at <unknown> (https://console-openshift-console.apps.pmarekipi.maistra.upshift.redhat.com/__cypress/runner/cypress_runner.js:119293:77)
      at tryCatcher (https://console-openshift-console.apps.pmarekipi.maistra.upshift.redhat.com/__cypress/runner/cypress_runner.js:1807:23)
      at Promise._settlePromiseFromHandler (https://console-openshift-console.apps.pmarekipi.maistra.upshift.redhat.com/__cypress/runner/cypress_runner.js:1519:31)
      at Promise._settlePromise (https://console-openshift-console.apps.pmarekipi.maistra.upshift.redhat.com/__cypress/runner/cypress_runner.js:1576:18)
      at Promise._settlePromise0 (https://console-openshift-console.apps.pmarekipi.maistra.upshift.redhat.com/__cypress/runner/cypress_runner.js:1621:10)
      at Promise._settlePromises (https://console-openshift-console.apps.pmarekipi.maistra.upshift.redhat.com/__cypress/runner/cypress_runner.js:1701:18)
      at _drainQueueStep (https://console-openshift-console.apps.pmarekipi.maistra.upshift.redhat.com/__cypress/runner/cypress_runner.js:2407:12)
      at _drainQueue (https://console-openshift-console.apps.pmarekipi.maistra.upshift.redhat.com/__cypress/runner/cypress_runner.js:2400:9)
      at Async._drainQueues (https://console-openshift-console.apps.pmarekipi.maistra.upshift.redhat.com/__cypress/runner/cypress_runner.js:2416:5)
      at Async.drainQueues (https://console-openshift-console.apps.pmarekipi.maistra.upshift.redhat.com/__cypress/runner/cypress_runner.js:2286:14)
  From Your Spec Code:
      at Context.eval (https://console-openshift-console.apps.pmarekipi.maistra.upshift.redhat.com/__cypress/tests?p=cypress/e2e/featureFiles/sidebar-navigation.feature:15941:34)

```

**Review by ScriptingShrimp — CHANGES_REQUESTED** [review]

File: plugin/cypress/support/step_definitions/hooks.ts
Comment (ferhoyos) [reply plugin/cypress/support/step_definitions/hooks.ts]: I have simplified the hook file removing tags that are not used in OSSMC tests and fixed a small issue in OSSMC creation yaml in manage-ossmc.sh script. Anyway these changes will not fix your issue. How have you installed OSSMC? Error says `kiali-operator` namespace is not found, you need the kiali operator installed to install OSSMC right?

Comment (ScriptingShrimp) [reply plugin/cypress/support/step_definitions/hooks.ts]: I do install OSSMC by executing Cypress, which do install OSSMC [via this script ](https://github.com/kiali/openshift-servicemesh-plugin/blob/f8e93a36a41d6d06579057398256e41392de62a5/hack/manage-ossmc.sh#L11). The yaml is copied from [Kiali website](https://kiali.io/docs/installation/installation-guide/creating-updating-ossmconsole-cr/#installing-via-oc-cli) how to install OSSMC, so if I understand this correctly the Kiali operator belongs into `openshift-operators` namespace. 

I tried to modify scripts to execute the cypress hooks in correct namespace, still no luck. Currently I'm provisioning fresh cluster to verify whole downstream pipeline - its possible I have something miss-configured in my environment

Comment (ferhoyos) [reply plugin/cypress/support/step_definitions/hooks.ts]: Yes, you are right. If you install Kiali operator from Openshift OperatorHub the Kiali operator belongs to `openshift-operators` namespace, not the `kiali-operator` one. In that case we will have to create specific `wait-for-namespace` script for OSSMC or modify the current one to include this case.

Comment (ScriptingShrimp) [reply plugin/cypress/support/step_definitions/hooks.ts]: I retested whole pipeline and I again see installed Kiali operator like: 
image

So there is deployment named `kiali-operator` which do belong into `openshift-operators` namespace

The `wait-for-namespace` and other hacks scripts works for normal regression, so I'm not exactly sure where is the difference, imho we should look into that before modyfing hack scripts?

**Comment by jmazzitelli** [comment]
I haven't read all of the notes here, but here are some things I think that will be helpful.

First, do not hardcode the namespace where you think the CR will be. It could be in `kiali-operator` if using our make targets to install things; or it could be in `openshift-operators` if things are installed via OLM. What you want your script to do is to find it dynamically. This is how you do it (I will assume you have one and only one Kiali CR installed):

```
KIALI_CR_NAMESPACE="$(oc get kiali --all-namespaces -o jsonpath='{..metadata.namespace}')"
```

That will output the namespace where the Kiali CR is located. If you do not know the name of the Kiali CR either, then you can do this to get the name and the namespace of the Kiali CR:

```
KIALI_CR_NAMESPACE_NAME="$(oc get kiali --all-namespaces -o custom-columns=NS:.metadata.namespace,N:.metadata.name --no-headers | sed 's/  */:/g')"
KIALI_CR_NAMESPACE="$(echo ${KIALI_CR_NAMESPACE_NAME} | cut -d: -f1)"
KIALI_CR_NAME="$(echo ${KIALI_CR_NAMESPACE_NAME} | cut -d: -f2)"
```

OK, at this point, you have the Kiali CR namespace (and I assume you know the name, too). Next you want to wait for the operator to finish reconciling the CR. We do something similar in our molecule tests as you see [here](https://github.com/kiali/kiali-operator/blob/v1.79.0/molecule/common/wait_for_kiali_cr_changes.yml#L12-L13)):

1. Get the reason and status for condition whose message is "Awaiting next reconciliation" (if there IS no condition whose message is "Awaiting next reconciliation" then the reconciliation is not done yet and you need to keep waiting):
KIALI_CR_REASON="$(oc get kiali $KIALI_CR_NAME -n $KIALI_CR_NAMESPACE -o jsonpath='{.status.conditions[?(@.message=="Awaiting next reconciliation")].reason}')"
KIALI_CR_STATUS="$(oc get kiali $KIALI_CR_NAME -n $KIALI_CR_NAMESPACE -o jsonpath='{.status.conditions[?(@.message=="Awaiting next reconciliation")].status}')"
2. If KIALI_CR_REASON is  "Successful" and KIALI_CR_STATUS is "True" then you can stop waiting - the operator has finished reconciling the CR. Otherwise keep waiting.

I am not sure if you can use those conditions in a "oc wait" command. I thought it was possible, but I can't find it in my notes - I looked into this a while ago. If I find it, I'll add another comment, but you can easily do the above in a script. Something like (I just tested this myself and it works):

```bash
KIALI_CR_NAMESPACE_NAME="$(oc get kiali --all-namespaces -o custom-columns=NS:.metadata.namespace,N:.metadata.name --no-headers | sed 's/  */:/g')"
KIALI_CR_NAMESPACE="$(echo ${KIALI_CR_NAMESPACE_NAME} | cut -d: -f1)"
KIALI_CR_NAME="$(echo ${KIALI_CR_NAMESPACE_NAME} | cut -d: -f2)"

echo -n "Waiting for operator to finish reconciling the CR named [$KIALI_CR_NAME] located in namespace [$KIALI_CR_NAMESPACE]"
while [ "$KIALI_CR_REASON" != "Successful" -o "$KIALI_CR_STATUS" != "True" ]; do
  sleep 1
  echo -n "."
  KIALI_CR_REASON="$(oc get kiali $KIALI_CR_NAME -n $KIALI_CR_NAMESPACE -o jsonpath='{.status.conditions[?(@.message=="Awaiting next reconciliation")].reason}')"
  KIALI_CR_STATUS="$(oc get kiali $KIALI_CR_NAME -n $KIALI_CR_NAMESPACE -o jsonpath='{.status.conditions[?(@.message=="Awaiting next reconciliation")].status}')"
done
echo
echo "Done reconciling"
```

File: plugin/cypress/support/step_definitions/hooks.ts
Comment (jmazzitelli) [reply plugin/cypress/support/step_definitions/hooks.ts]: FYI: by default, the Kiali Server will be installed in the same namespace where the Kiali CR is. If you are running in an OLM environment where the operator is installed in "openshift-operators" and you create the CR via the OpenShift Console GUI, the Kiali CR will be created in "openshift-operators". By default, therefore, the Kiali Server will be installed in that namespace, too. If you do not want that behavior, override the default by explicitly setting `spec.deployment.namespace` in the Kiali CR. For the docs on this setting, see: https://kiali.io/docs/configuration/kialis.kiali.io/#.spec.deployment.namespace
```
.spec.deployment.namespace
The namespace into which Kiali is to be installed. If this is empty or not defined, the default will be the namespace where the Kiali CR is located.
```

File: hack/wait-for-namespace.sh
Comment (jmazzitelli) [inline hack/wait-for-namespace.sh:48]: I think you mean the CR ("Kiali CR") -- the CRD is the kialis.kiali.io CustomResourceDefinition.

File: hack/wait-for-namespace.sh
Comment (jmazzitelli) [inline hack/wait-for-namespace.sh:71]: Don't understand why wait is done twice here (?).

**Review by jmazzitelli — APPROVED** [review]
I didn't run any of this, but I assume it is doing what you need it to do.

File: hack/wait-for-namespace.sh
Comment (ferhoyos) [reply hack/wait-for-namespace.sh:71]: Not sure, this part is copied from `wait-for-namespace.sh` script in Kiali repo, made by @matejnesuta in https://www.github.com/kiali/kiali/pull/6257. Do you remember why wait is done twice here? Probably there is a good reason to do that.

File: hack/wait-for-namespace.sh
Comment (ferhoyos) [reply hack/wait-for-namespace.sh]: Oops, yes I mean CR. Typo fixed

File: hack/wait-for-namespace.sh
Comment (matejnesuta) [reply hack/wait-for-namespace.sh:71]: Hello Fernando. The code in the script originated from one of our Jenkins jobs, which is used for installing the demoapps. I had to re-use this piece of code in my Cypress hook, otherwise it would not work on our OCP pipelines, so I decided to create this little script and it is now executed both by the hook and by the Jenkins pipeline. 

Frankly, I am also not sure why the line is present twice here.

Comment (ferhoyos) [reply hack/wait-for-namespace.sh]: Ok, in that case I will remove the `oc wait` command duplicated

Comment (jmazzitelli) [reply hack/wait-for-namespace.sh]: I see this code at the end of this script:

```bash
for NAMESPACE in ${NAMESPACES[@]}; do
  oc wait --for=condition=Ready pods --all -n "$NAMESPACE" --timeout 60s
done

sleep 80
```

It looks like this is simply waiting for everything to start (i.e. it waits for all pods in each namespace to be ready, then it waits another 80 seconds before exiting).

Notice the line `oc wait` has a timeout of 60s. If 60 seconds passes and all the pods are not ready, the `oc wait` will exit with a non-zero exit code. This will abort the script (due to `set -e` on line 6, which tells the script to exit on any command that results in an error). That means it will not wait for any more namespaces to be ready and it will not wait that extra 80 seconds.

Now, notice the line you removed in the [last commit](https://github.com/kiali/openshift-servicemesh-plugin/pull/245/commits/8a92d07ebfa45ea085994dbb05a4771714460a9b) had `|| true` at the end - this means even if the `oc wait` timed out, it will not cause the script to abort. So whoever wrote that code originally must have wanted that code to keep waiting even if one namespace isn't ready (i.e. he didn't want the script to abort if `oc wait` timed out) - he wanted to wait for the other namespaces and that extra 80 seconds even if one namespace wasn't ready after 60s. I guess this is to wait as long as possible so the test components have a chance to start fully before the tests begin to run.

I will leave it up to you if you want to maintain that same behavior as before and put back that `|| true` or if you want to leave it out. I'll approve this PR in case you want to leave it as-is and you can merge.

**Review by jmazzitelli — DISMISSED** [review]

File: hack/wait-for-namespace.sh
Comment (ferhoyos) [reply hack/wait-for-namespace.sh]: Good catch @jmazzitelli. Eventually this script is used for cypress test configuration, so I prefer to set `|| true` and don't abort the test process if the pods are not ready (to avoid failures if the system is quite slow that day). If they are not ready after the extra 80s wait then cypress tests will fail afterwards. @matejnesuta what do you think?

**Review by jmazzitelli — APPROVED** [review]

File: hack/wait-for-namespace.sh
Comment (matejnesuta) [reply hack/wait-for-namespace.sh]: I have no problems removing one of the `oc wait` commands as long as it works.

**Comment by ScriptingShrimp** [comment]
testing this on OCP, I cannot see bookinfo in service mesh, nor in Kiali. Execution of the scripts passes, but the test fail on the visual assertion in the dropdown on the image below

image

File: hack/wait-for-namespace.sh
Comment (nrfox) [inline hack/wait-for-namespace.sh:41]: You can ignore this but it's often easier to use jsonpath and/or jq to parse this stuff than it is to chop up the text e.g. `oc get kiali --all-namespaces -o jsonpath='{.items[*].metadata.namespace}'`

File: hack/wait-for-namespace.sh
Comment (ferhoyos) [reply hack/wait-for-namespace.sh:1]: Thanks for catching up this @nrfox. I add this script manually because I added some changes here. But I think the best approach is to include these changes into Kiali repo and then download it from there. 

Kiali PR to include script changes: https://github.com/kiali/kiali/pull/7230

**Comment by ScriptingShrimp** [comment]
Hi @ferhoyos , I'm still getting error related to bookinfo on OCP
```
Kiali sidebar integration with OCP Console
    ✓ Service mesh buttons are displayed (17644ms)
    ✓ Overview page is displayed correctly (8283ms)
    1) Graph page is displayed correctly
    ✓ Istio Config page is displayed correctly (9737ms)


  3 passing (8m)
  1 failing

  1) Kiali sidebar integration with OCP Console
       Graph page is displayed correctly:
     CypressError: `cy.exec('../_output/kiali/hack/istio/wait-for-namespace.sh -n bookinfo')` timed out after waiting `400000ms`.
```

were you able to run this locally? 
I'll investigate more in detail tomorrow .

**Review by ScriptingShrimp — APPROVED** [review]
I will merge this now, lets see if this passing on CI. 
If not we will revert the merge commit on Monday.

**Review by nrfox — APPROVED** [review]

**Pull request merged**
